Allow replicator documents to include params for db creation

jiangphcn opened the pull request:
<!-- Thank you for your contribution!

     Please file this form by replacing the Markdown comments
     with your text. If a section needs no action - remove it.

     Also remember, that CouchDB uses the Review-Then-Commit (RTC) model
     of code collaboration. Positive feedback is represented +1 from committers
     and negative is a -1. The -1 also means veto, and needs to be addressed
     to proceed. Once there are no objections, the PR can be merged by a
     CouchDB committer.

     See: http://couchdb.apache.org/bylaws.html#decisions for more info. -->

## Overview

<!-- Please give a short brief for the pull request,
     what problem it solves or how it makes things better. -->
If you specify "create_target": true in a replicator doc, the replicator creates the target database, with the default q value.

With this PR, it is allowed to specify "create_database_params": {"q": "1"} to have the replicator to create the target database with specified value (currently, the example is q=1 and q can be other value). The example of replication document looks like:
```
{
  "source": "http://foo:bar@localhost:15984/db1",
  "target": "http://foo:bar@localhost:15984/db22",
  "create_target": true,
  "continuous": false,
   "create_target_params": {"q": "1"}

}
```

One thing to be highlighted is about the value in `source` and `target` field in replication document.

Most of case database in `url` format were specified in replication document. 
```
{
  "source": "http://foo:bar@localhost:15984/db1",
  "target": "http://foo:bar@localhost:15984/db2",
  "create_target": true,
  "continuous": true
}
```
However, I found that the format `"source": "shards/00000000-1fffffff/db3.1511408454"` was also supported in replication document. But the database should be one in shard instead of logical database name. It looks to me that it doesn't make sense to specify q value if the shard database was specified.
https://github.com/apache/couchdb/pull/1017/files#diff-3a45ba156c61480ff98a83509af1b2cdR135

So right now, only database creation with url format approached was extended.

## Testing recommendations

<!-- Describe how we can test your changes.
     Does it provides any behaviour that the end users
     could notice? -->
```
make check skip_deps+=couch_epi apps=couch_replicator tests=create_target_with_options_replication_test_

    Running test function(s):
      couch_replicator_create_target_with_options_tests:create_target_with_options_replication_test_/0
======================== EUnit ========================
Create target with range partitions tests
  couch_replicator_create_target_with_options_tests:63: should_create_target_with_q_4...ok
  couch_replicator_create_target_with_options_tests:63: should_create_target_with_q_4...ok
  couch_replicator_create_target_with_options_tests:84: should_create_target_with_q_2_n_1...ok
  couch_replicator_create_target_with_options_tests:85: should_create_target_with_q_2_n_1...ok
  couch_replicator_create_target_with_options_tests:84: should_create_target_with_q_2_n_1...ok
  couch_replicator_create_target_with_options_tests:85: should_create_target_with_q_2_n_1...ok
  couch_replicator_create_target_with_options_tests:105: should_create_target_with_default...ok
  couch_replicator_create_target_with_options_tests:105: should_create_target_with_default...ok
  couch_replicator_create_target_with_options_tests:121: should_not_create_target_with_q_any...ok
  couch_replicator_create_target_with_options_tests:121: should_not_create_target_with_q_any...ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
  [done in 6.025 s]
=======================================================
  All 10 tests passed.
==> rel (eunit)
```
## Related Issues or Pull Requests

<!-- If your changes affects multiple components in different
     repositories please put links to those issues or pull requests here.  -->
https://github.com/apache/couchdb/issues/887
## Checklist

- [X] Code is written and works correctly;
- [X] Changes are covered by tests;
- [ ] Documentation reflects the changes;

nickva commented on src/couch_replicator/src/couch_replicator_api_wrap.erl:
```diff
@@ -68,20 +68,24 @@ db_uri(Db) ->
             ok;
         true ->
-            send_req(Db, [{method, put}],
+            Db2 = case couch_util:get_value(<<"q">>, CreateParams) of
```
We explicitly handle `q` parameter here, I wonder if it is worth making it more generic. What if users want to specify `n`, `placement` or even the new PSE `engine` parameters? I think it should be possible. Or do you think it's better to restrict it to specific params only?

jiangphcn replied:
Yes, I extend this to be generic.

nickva commented:
@jiangphcn Good work. Thanks for taking initiative and giving it a try.

See some comments about tests and a general comment about the approach taken. I saw that q is special-cased and we don't allow other parameters. Wonder if we want to make it generic and allow users to specify other query parameters as well (n, placement, engine etc). In that case don't forget to also urlencode stuff, lookup query parameters encoding rules.

As for local dbs, those are not just `shards/...` dbs could also be regular (simple) node local dbs, say 'my_target_db`. This endpoint format in general is problematic and doesn't work as most users expect it. It used to create local dbs in a CouchDB 1.x and 2.x with clustering it should really create a db via fabric, now it just creates a db on the random node where replication runs. Ideally we might want to handle some parameters there as well (the ones that make sense, like say `engine` for future PSE work). But in practice, I think we shouldn't bother and just ignore it params like you already do.

jiangphcn commented on src/couch_replicator/test/couch_replicator_create_target_with_options_tests.erl:
```diff
@@ -0,0 +1,168 @@
+    ?_assertEqual(list_to_integer(Q), couch_util:get_value(q, ClusterInfo)).
+
+
+should_not_create_target_with_q_any(
```
This test case is used to prevent from regression. If `create_target` is not specified or set to false, there should not be target database to be created, even if there is any value set for `create_target_params`.

jiangphcn commented on src/couch_replicator/test/couch_replicator_create_target_with_options_tests.erl:
```diff
@@ -0,0 +1,168 @@
+            foreachx,
+            fun setup/1, fun teardown/2,
+            [{P, fun should_create_target_with_q_1/2} || P <- Ps] ++
+            [{P, fun should_create_target_with_q_4/2} || P <- Ps] ++
```
Okay, keep `q=4`, and also add `q_2_n_1`.

jiangphcn commented:
@nickva Thanks a lot for review and comments. I addressed your comment using new commit d4e9230, mainly make parameter of creating target generic and adapt test cases. Please help review again. Thanks

nickva commented on src/couch_replicator/src/couch_replicator_api_wrap.erl:
```diff
@@ -68,20 +68,25 @@ db_uri(Db) ->
             ok;
         true ->
-            send_req(Db, [{method, put}],
+            Db2 = if CreateParams == [] -> Db; true ->
```
Let's pull this into a separate function and handle the CreateParams=[] in the head.

`Db2 = maybe_append_create_query_params(Db, CreateParams)` ?
